### DISCUSSION TITLE
How to prevent multiple nodes with same PeerId?

### DISCUSSION THREAD

**Discussion by erhant** [libp2p/rust-libp2p#5641] [Q&A]
Hi, in our network (https://github.com/firstbatchxyz/dkn-compute-node) we currently have the issue of people running multiple nodes using the same private key, and therefore having the same PeerId in different machines. We would like to prevent that, but could not yet see how to do it. Would love some ideas 🙏🏻 

I have tried a simple check within `identify::Event::Received` handler, such as:
```rs
if peer_id == *self.swarm.local_peer_id() {
  panic!("Identify: Received self identify event, are you running multiple nodes with same wallet?");
}
```

but that did not catch the node with same PeerId.

For additional context, [here](https://github.com/firstbatchxyz/dkn-compute-node/blob/master/p2p/src/client.rs#L259) is where we handle the identify event, and add the identified peer to our table.

**Comment by guillaumemichel** [comment]
I don't understand why nodes would share their private keys with one another.

If you are using kademlia you can always try to lookup the newly generated identity when joining the network. If the peerid already exist, generate a fresh id.

**Reply by erhant** [reply]
> I don't understand why nodes would share their private keys with one another.

In our case, people are running local LLM models to do tasks, and they get rewards for it based on their signatures from which we derive their Ethereum address and such. Since one machine can only fit so many LLM models at once, but people want to have the rewards to a single wallet only for convenience, they use the same private key and spin up multiple machines.

> If you are using kademlia you can always try to lookup the newly generated identity when joining the network. If the peerid already exist, generate a fresh id.

We do use Kademlia, I will see what we can do 👍🏻 Im more on the side of: "if the peerid exists, tell the user to use a differnet private key and run the node again"

**Reply by erhant** [reply]
I have tried to handle the `get_closest_peers` result with:

```rs
   /// Handles the results of a Kademlia closest peers search, simply logs it.
    fn handle_closest_peers_result(
        &mut self,
        result: Result<GetClosestPeersOk, GetClosestPeersError>,
    ) {
        match result {
            Ok(GetClosestPeersOk { peers, .. }) => {
                if peers
                    .iter()
                    .find(|p| p.peer_id == *self.swarm.local_peer_id())
                    .is_some()
                {
                    panic!("Kademlia: Found self in closest peers.");
                }
                log::info!(
                    "Kademlia: Query finished with {} closest peers.",
                    peers.len()
                );
            }
            Err(GetClosestPeersError::Timeout { peers, .. }) => {
                log::info!(
                    "Kademlia: Query timed out with {} closest peers.",
                    peers.len()
                );
            }
        }
    }
```

You can imagine that the event result is passed to this function directly, and `self.swarm` is the Swarm instance generated with the peerId that belongs to the node. I would expect the `find(|p| p.peer_id == *self.swarm.local_peer_id())` check to catch this but so far it did not, does this look correct to you?

**Reply by erhant** [reply]
P.S. sorry to the belated answer, I wrote a reply originally as a different answer thinking the discussion page is like an issue page :D

**Reply by guillaumemichel** [reply]
I realize it is a bit trickier than I expected

If a node tries to find DHT records for its own peer id, but not matching itself, it must check whether the multiaddresses of the obtained record (matching its own peerid) match its own multiaddresses. If not, then another node is using the same peerid.

Alternatively, it is certainly easier to generate a fresh identity, make the DHT lookup for the peerid corresponding to the private key that the node wants to use. If no DHT record is found, the node identity can safely be switched to the private key. Otherwise, the private key is already in use.

**Reply by guillaumemichel** [reply]
And also, the above only work if the other node that may be using the specific private key is advertising itself as a DHT server. Some nodes are client only, and they are harder to detect

**Reply by erhant** [reply]
Hmm, thankfully we are in Server mode (https://github.com/firstbatchxyz/dkn-compute-node/blob/master/p2p/src/client.rs#L92 to confirm). I can check the multiaddr indeed, will try! Thank you for the quick reply.

**Reply by erhant** [reply]
Where should I handle this exactly though?

**Reply by guillaumemichel** [reply]
I would say generating a new identity for the DHT lookup and using the private key only if the DHT lookup didn't find any peer using the private key is probably the easier and safer approach.

**Reply by erhant** [reply]
I'm still unsure how to check my own identity within DHT. Even when I make a request to  `swarm.behaviour_mut().kademlia.get_closest_peers` with my own peer id, I cant find myself in the emitted events.

For event, im handing it as:

```rs
match self.swarm.select_next_some().await {
                SwarmEvent::Behaviour(DriaBehaviourEvent::Kademlia(
                    kad::Event::OutboundQueryProgressed {
                        result: QueryResult::GetClosestPeers(result),
                        ..
                    },
                )) => self.handle_closest_peers_result(result),
// ...
```

where `handle_closest_peers_result` is:

```rs
 fn handle_closest_peers_result(
        &mut self,
        result: Result<GetClosestPeersOk, GetClosestPeersError>,
    ) {
        match result {
            Ok(GetClosestPeersOk { peers, .. }) => {
                // log stuff, cant find my own peer id in `peers`
            }
            Err(GetClosestPeersError::Timeout { peers, .. }) => {
                // log stuff here
            }
        }
    }
```   

Am I doing something wrong?

**Reply by guillaumemichel** [reply]
IIRC `kad` is filtering out results about self, so this technique won't work indeed.

I suggest to generate a random identity to make the `kad` lookup about the identity/privkey you are interested in. If you don't find it in the DHT, you can use it safely.
